### Loading, exploring and validating the data

In [7]:
# Function to fetch data from the NYC taxi data website

from pathlib import Path
%pip install requests
import requests

def download_one_file_of_raw_data(year: int, month: int) ->Path:
    """Download one file of raw data from the NYC taxi data website."""

    url = f"https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_{year}-{month:02}.parquet"
    response = requests.get(url)

    if response.status_code == 200:
        path = f"../data/raw/rides_{year}_{month:02}.parquet"
        open(path, 'wb').write(response.content)
        return path
    else:   
        raise RuntimeError(f"Failed to download {url}")
    

  Obtaining dependency information for requests from https://files.pythonhosted.org/packages/70/8e/0e2d847013cb52cd35b38c009bb167a1a26b2ce6cd6965bf26b47bc0bf44/requests-2.31.0-py3-none-any.whl.metadata
  Obtaining dependency information for charset-normalizer<4,>=2 from https://files.pythonhosted.org/packages/57/ec/80c8d48ac8b1741d5b963797b7c0c869335619e13d4744ca2f67fc11c6fc/charset_normalizer-3.3.2-cp311-cp311-win_amd64.whl.metadata
  Obtaining dependency information for idna<4,>=2.5 from https://files.pythonhosted.org/packages/c2/e7/a82b05cf63a603df6e68d59ae6a68bf5064484a0718ea5033660af4b54a9/idna-3.6-py3-none-any.whl.metadata
  Obtaining dependency information for urllib3<3,>=1.21.1 from https://files.pythonhosted.org/packages/96/94/c31f58c7a7f470d5665935262ebd7455c7e4c7782eb525658d3dbf4b9403/urllib3-2.1.0-py3-none-any.whl.metadata
  Obtaining dependency information for certifi>=2017.4.17 from https://files.pythonhosted.org/packages/64/62/428ef076be88fa93716b576e4a01f919d25968913e81


[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: C:\Users\User\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [8]:
# Call the function to download the data for January 2022
download_one_file_of_raw_data(2022,1)

'../data/raw/rides_2022_01.parquet'

In [10]:
%pip install pandas # Install pandas

  Obtaining dependency information for pandas from https://files.pythonhosted.org/packages/11/17/fb1a34f3e73debbc2fd15a01ea17eaab3717943d08463ff4979a4f024b3f/pandas-2.1.4-cp311-cp311-win_amd64.whl.metadata
  Obtaining dependency information for numpy<2,>=1.23.2 from https://files.pythonhosted.org/packages/99/2b/f7114983d84303019385d93d24d729aedba67be7e083286f114188943cf3/numpy-1.26.3-cp311-cp311-win_amd64.whl.metadata
     ---------------------------------------- 0.0/61.2 kB ? eta -:--:--
     ------------ ------------------------- 20.5/61.2 kB 682.7 kB/s eta 0:00:01
     ------------------------------- ------ 51.2/61.2 kB 660.6 kB/s eta 0:00:01
     -------------------------------------- 61.2/61.2 kB 653.5 kB/s eta 0:00:00
  Obtaining dependency information for pytz>=2020.1 from https://files.pythonhosted.org/packages/32/4d/aaf7eff5deb402fd9a24a1449a8119f00d74ae9c2efa79f8ef9994261fc2/pytz-2023.3.post1-py2.py3-none-any.whl.metadata
  Obtaining dependency information for tzdata>=2022.1 


[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: C:\Users\User\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [20]:
import pandas as pd # Import pandas

rides = pd.read_parquet('../data/raw/rides_2022_01.parquet') # Read the data into a pandas dataframe

rides.head(10) # Show the first 10 rows of the dataframe


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1         1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2         2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3         2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4         2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
5         1  2022-01-01 00:40:15   2022-01-01 01:09:48              1.0   
6         2  2022-01-01 00:20:50   2022-01-01 00:34:58              1.0   
7         2  2022-01-01 00:13:04   2022-01-01 00:22:45              1.0   
8         2  2022-01-01 00:30:02   2022-01-01 00:44:49              1.0   
9         2  2022-01-01 00:48:52   2022-01-01 00:53:28              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.80         1.0                  N           142           236   
1           2.10         1.0                  N           236            42   
2           0.97         1.0                  N           166           166   
3           1.09         1.0                  N           114            68   
4           4.30         1.0                  N            68           163   
5          10.30         1.0                  N           138           161   
6           5.07         1.0                  N           233            87   
7           2.02         1.0                  N           238           152   
8           2.71         1.0                  N           166           236   
9           0.78         1.0                  N           236           141   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         14.5    3.0      0.5        3.65          0.00   
1             1          8.0    0.5      0.5        4.00          0.00   
2             1          7.5    0.5      0.5        1.76          0.00   
3             2          8.0    0.5      0.5        0.00          0.00   
4             1         23.5    0.5      0.5        3.00          0.00   
5             1         33.0    3.0      0.5       13.00          6.55   
6             1         17.0    0.5      0.5        5.20          0.00   
7             2          9.0    0.5      0.5        0.00          0.00   
8             1         12.0    0.5      0.5        2.25          0.00   
9             2          5.0    0.5      0.5        0.00          0.00   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    0.3         21.95                   2.5          0.0  
1                    0.3         13.30                   0.0          0.0  
2                    0.3         10.56                   0.0          0.0  
3                    0.3         11.80                   2.5          0.0  
4                    0.3         30.30                   2.5          0.0  
5                    0.3         56.35                   2.5          0.0  
6                    0.3         26.00                   2.5          0.0  
7                    0.3         12.80                   2.5          0.0  
8                    0.3         18.05                   2.5          0.0  
9                    0.3          8.80                   2.5          0.0

In [21]:
# Shape of the dataframe
rides.shape

(2463931, 19)

In [22]:
# Info about the dataframe
rides.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2463931 entries, 0 to 2463930
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[ns]
 2   tpep_dropoff_datetime  datetime64[ns]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee           

**Observations and insights**

The data has 2,463,931 rows and 19 columns (January data), we may use more data in the future.

Majority of the columns are retundant for our purposes. Main columns of interest are 'tpep_pickup_datetime','tpep_dropoff_datetime','PULocationID'

We will need to generate new features through data engineering for further analysis and to improve our model in the future.

In [23]:
# checking for missing values
rides.isnull().sum() / len(rides) * 100

VendorID                 0.000000
tpep_pickup_datetime     0.000000
tpep_dropoff_datetime    0.000000
passenger_count          2.901989
trip_distance            0.000000
RatecodeID               2.901989
store_and_fwd_flag       2.901989
PULocationID             0.000000
DOLocationID             0.000000
payment_type             0.000000
fare_amount              0.000000
extra                    0.000000
mta_tax                  0.000000
tip_amount               0.000000
tolls_amount             0.000000
improvement_surcharge    0.000000
total_amount             0.000000
congestion_surcharge     2.901989
airport_fee              2.901989
dtype: float64

**Observations and insights**

There are a few columns with a low number of missing value (less than 3%).

As number of missing values are low, these won't affect our analysis in a meaningfull way.

Also the columns we will use for our projects do not have missing values, there is no for missing value imputation in this case.

In [24]:
# selecting the columns of interest for initial analysis
rides = rides[['tpep_pickup_datetime', 'PULocationID']] 

# renaming the columns for easier use
rides.rename(columns={
    'tpep_pickup_datetime': 'pickup_datetime',
    'PULocationID': 'pickup_location_id',
}, inplace=True)

rides.head() # show first 5 rows

pickup_datetime  pickup_location_id
0 2022-01-01 00:35:40                 142
1 2022-01-01 00:33:43                 236
2 2022-01-01 00:53:21                 166
3 2022-01-01 00:25:21                 114
4 2022-01-01 00:36:48                  68

**Conducting some sanity checks**

Lets conduct some can sanity checks to see if all loaded data is from January

In [26]:

rides['pickup_datetime'].describe() # show summary statistics for the pickup_datetime column

C:\Users\User\AppData\Local\Temp\ipykernel_28608\3581647394.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  rides['pickup_datetime'].describe() # show summary statistics for the pickup_datetime column


count                 2463931
unique                1423522
top       2022-01-26 07:57:00
freq                       12
first     2008-12-31 22:23:09
last      2022-05-18 20:41:57
Name: pickup_datetime, dtype: object

**Observations and insights**

We observe that there is a row from 2008 and a row from May 2022, let us fix this! we will take out all rows not from January 2022

In [27]:
rides = rides[rides.pickup_datetime >= '2022-01-01']
rides = rides[rides.pickup_datetime < '2022-02-01']
rides['pickup_datetime'].describe()

C:\Users\User\AppData\Local\Temp\ipykernel_28608\124399977.py:3: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  rides['pickup_datetime'].describe()


count                 2463879
unique                1423471
top       2022-01-26 07:57:00
freq                       12
first     2022-01-01 00:00:08
last      2022-01-31 23:59:58
Name: pickup_datetime, dtype: object

Let us save our validated data for future use.

In [6]:
rides.to_parquet('../data/transformed/validated_rides_2022_01.parquet') # save to parquet file